src/rank_llm/api/server.py

```diff
@@ -0,0 +1,108 @@
+import argparse
+from flask import Flask, jsonify, request
+
+from rank_llm import retrieve_and_rerank
+from rank_llm.rerank.rank_listwise_os_llm import RankListwiseOSLLM
+from rank_llm.rerank.api_keys import get_openai_api_key, get_azure_openai_args
+from rank_llm.rerank.rank_gpt import SafeOpenai
+from rank_llm.rerank.rankllm import PromptMode
+
+""" API URL FORMAT
+
+http://localhost:8082/api/model/{model_name}/index/{index_name}/{retriever_base_host}?query={query}&hits_retriever={top_k_retriever}&hits_reranker={top_k_reranker}&qid={qid}&num_passes={num_passes}
+
+hits_retriever, hits_reranker, qid, and num_passes are OPTIONAL
+Default to 20, 5, None, and 1 respectively
+
+"""
+ 
+
+def create_app(model, port, use_azure_openai=False):
+
+    app = Flask(__name__)
+    if model == 'rank_zephyr':
+        print(f"Loading {model} model...")
+        # Load specified model upon server initialization
+        default_agent = RankListwiseOSLLM(
+            model=f"castorini/{model}_7b_v1_full",
+            context_size=4096,
+            prompt_mode=PromptMode.RANK_GPT,
+            num_few_shot_examples=0,
+            device="cuda",
+            num_gpus=1,
+            variable_passages=True,
+            window_size=20,
+            system_message="You are RankLLM, an intelligent assistant that can rank passages based on their relevancy to the query.",
+        )
+    elif model == 'rank_vicuna':
+        print(f"Loading {model} model...")
+        # Load specified model upon server initialization
+        default_agent = RankListwiseOSLLM(
+            model=f"castorini/{model}_7b_v1",
+            context_size=4096,
+            prompt_mode=PromptMode.RANK_GPT,
+            num_few_shot_examples=0,
+            device="cuda",
+            num_gpus=1,
+            variable_passages=False,
+            window_size=20,
+        )
+    elif 'gpt' in model:
+        print(f"Loading {model} model...")
+        # Load specified model upon server initialization
+        openai_keys = get_openai_api_key()
+        print(openai_keys)
+        default_agent = SafeOpenai(
+            model=model,
+            context_size=8192,
+            prompt_mode=PromptMode.RANK_GPT,
+            num_few_shot_examples=0,
+            keys=openai_keys,
+            **(get_azure_openai_args() if use_azure_openai else {}),
+        )
+    else:
+        raise ValueError(f"Unsupported model: {model}")
+
+    @app.route('/api/model/<string:model_path>/index/<string:dataset>/<string:host>', methods=['GET'])
+    def search(model_path, dataset, host):
+
+        query = request.args.get('query',type=str)
+        top_k_retrieve = request.args.get('hits_retriever',default=20,type=int)
+        top_k_rerank = request.args.get('hits_reranker',default=5,type=int)
+        qid = request.args.get('qid',default=None,type=str)
+        num_passes = request.args.get('num_passes',default=1,type=int)
```

Comment:
These lines seem to need formatting, please install and run black for formatting.